qiskit/dagcircuit/dagcircuit.py

```diff
@@ -262,25 +262,28 @@ def apply_operation_back(self, op, qargs=None, cargs=None, condition=None):
         self._check_bits(qargs, self.output_map)
         self._check_bits(all_cbits, self.output_map)
 
-        self._add_op_node(op, qargs, cargs, condition)
+        node_index = self._add_op_node(op, qargs, cargs, condition)
 
         # Add new in-edges from predecessors of the output nodes to the
         # operation node while deleting the old in-edges of the output nodes
         # and adding new edges from the operation node to each output node
         al = [qargs, all_cbits]
         for q in itertools.chain(*al):
-            ie = list(self._multi_graph.predecessors(self.output_map[q]))
+            ie = list(self._multi_graph.predecessors(self.output_map[q]._node_id))
```

Comment:
I like that idea. We rely more on node id now because retworkx is index addressed so having a public attribute with less characters would be nice. Were you thinking of a property/setter pair?

Reply:
I was just forgetting about setter (`node._node_id = 123` was in mind) but providing setter would be nice. I have no preference. It's up to you.

Reply:
In my opinion, `node_id` shouldn't be part of the `DAGNode` API and in the long term, shouldn't be tracked by `DAGNode` instances. The prior implementation had taken steps toward removing `node_id` as it had led to confusion (incorrect use as a proxy for things like graph ordering, instances where a `DAGNode` is a member of more than one `DAGCircuit`).

Instead, `DAGCircuit` could track a bidirectional `[Int, DAGNode]` map. This doesn't need to happen here, but I'd advocate against making `node_id` public.

Reply:
That's fine, I don't really have a strong opinion on it. I'll leave it as is for now. FWIW, when I first started this PR I had `_id_to_node` dict and a `_node_to_id` dict. But, it got to be pretty verbose adding things like `self._multi_graph.predecessors(self._node_to_id[self.output_map[q]])` all over the place, which is why I settled on using the `_node_id` attribute in the dagnode objects